src/Lumi/Components/Form.purs

```diff
@@ -254,7 +255,8 @@ useForm editor props = Hooks.do
 -- | to be provided as an additional argument.
 useForm'
   :: forall ui props unvalidated result
-   . FormBuilder' ui props unvalidated result
+   . Mapping Validation.ModifyValidated unvalidated unvalidated
```

Comment:
Yeah, this is because in the instance chain for `Mapping ModifyValidated` we have an identity base case that's valid for any type `a`. So if we don't specify this constraint it ends up always picking that identity instance :/

I don't know how to force the compiler require this constraint whenever we use `setModified`. Maybe @hdgarrood would know?

Reply:
I don't think so. I think this is related to https://github.com/purescript/purescript/issues/3531, it might be a bug in the compiler that it compiled previously.

Reply:
Having said that, I'm not completely convinced that all of the examples listed in that issue are actually behaving incorrectly at the moment. It also seems a bit dodgy that the compiler can differentiate between a concrete type and a type variable when deciding whether a catch-all instance like the one at the end of that chain applies. I really just need to reread the instance chains paper, because I don't fully understand them right now.

Reply:
It does seem fishy that that instance can be called without knowing either 1) the specific type, or 2) having the constraint propagated up until something gives it a specific type. For example, the react hooks library works on the assumption that the type has to be decided at some point -- it doesn't matter what the hook type is, as long as there is one.

Another example:

```purs
case _ of
  Just a -> ?a
  _ -> ?b
```

This should obviously be `Maybe a -> b`, because of the `Just`. But the fallback instance here is kind of like the the compiler inferring `f c -> b` for this `case` because the 2nd path is a wildcard to catch non-maybe values. Not sure if that makes sense..

Reply:
I know neither of these examples are classes or instances, but it seemed like the same principle. Maybe classes are different.

Reply:
In general, type variables don't always have to be 'eventually concrete'. For example, no concrete type is ever selected for the Region parameter of ST when you do `runST`. For a simpler example, if I write
```
thing :: Int
thing = const 0 (unsafeCoerce unit)
```
that's fine; we don't ever make the type of `unsafeCoerce unit` concrete. I don't follow your Maybe example, though.

Reply:
> no concrete type is ever selected for the Region parameter of ST when you do `runST`

Interesting.. I thought it had to decide one. Is this related to errors about escaped skolems? Would that be the error here if the compiler didn't allow choosing this instance using an undecided type?

> I don't follow your Maybe example

I was trying to compare to code I'm more familiar with -- the identity instance here ends up being so general the compiler lets you put anything you want there, and I was wondering if you thought it was a bug for a similar reason to that `case` example, in the same way that generalizing away the `Nothing` could technically still produce valid code, even though it doesn't make sense.

Reply:
I guess I was trying to ask if there is a valid use case for selecting an instance using unknown types

Reply:
It's sort of related to errors about escaped skolems, but I don't think we would see that error here. I think you only see escaped skolem errors with higher rank types, like `runST :: forall a. (forall r. ST r a) -> a`, i.e. when you have a `forall` inside an argument; if all of the foralls come right at the start of the type signature, then their scopes encompass the whole type, so there's nowhere to escape to, I think.

I think in general, there are valid use cases for selecting instances using unknown types. For example, there is an instance `Monoid a => Monoid (r -> a)`. If I write `maybe "hi" mempty Nothing`, then I know that `mempty` should be a function which takes whatever is contained within the last argument to `maybe`. Here, that's `Nothing`, which means that we don't know what that type is. But that's fine, because we know enough to commit to the instance; I would expect this to compile without me having to annotate it like `(Nothing :: Maybe Int)`.

In my mind, it's really just instance chains which frustrate things. In the case of this PR, or just in general when you bring instance chains in, I no longer know how things are supposed to work.

Reply:
Ah, that makes sense, thank you!